File: python/389.找不同.py
#
# @lc app=leetcode.cn id=389 lang=python
#
# [389] 找不同
#
# 解法1(T70% S42%)：题目没说清楚，每个字符可能出现多次，因此要用字典不能只用集合。先一次遍历s构建字典，然后一次遍历t，如果某个字符不在字典中或个数已经减为0了就返回

# @lc code=start
class Solution(object):
    def findTheDifference(self, s, t):
        """
        :type s: str
        :type t: str
        :rtype: str
        """
        table = {}
        for l in s:
            table[l] = table.get(l, 0) + 1

        for l in t:
            if l not in table or table[l] == 0: return l
            else: table[l] -= 1
# @lc code=end

File: javascript/389.找不同.js
/*
 * @lc app=leetcode.cn id=389 lang=javascript
 *
 * [389] 找不同
 * 
 * js对集合的支持也很差
 */

// @lc code=start
/**
 * @param {string} s
 * @param {string} t
 * @return {character}
 */
var findTheDifference = function(s, t) {
    s = s.split("").sort();
    t = t.split("").sort();
    for (var i=0,j=0; i<s.length && j<t.length; i++,j++) {
        if (s[i]!=t[j]) {
            return t[j];
        }
    }
    return t[t.length-1];
};
// @lc code=end

